src/dataflow/source/kafka.rs

```diff
@@ -462,6 +468,11 @@ fn downgrade_capability(
                     );
                     needs_md_refresh = true;
                 }
+                // This assertion makes sure that if we ever fast-forwarded the empty stream
+                // (any message of the form "PID,TS,0"), we have correctly removed the (_,_,0) entry
+                // even if data has subsequently been added to the stream
+                assert!(*offset != 0 || last_offset == 0);
```

Comment:
Should it always be true that `offset >= last_offset` ? That could be an even more comprehensive assert.